### PR TITLE
chore: Add psalm for code analysis

### PR DESCRIPTION
**Pull request by susnux**
Fixes https://github.com/nextcloud/suspicious_login/issues/705

Fix all detected problem.
There should be no¹ change in logic or behavior.

¹ Well one exception: If a suspicious login could not be persisted the code threw at some random place, so instead we handle this by notify the user and log that as critical severity.

### CODE REVIEW

File: lib/Service/LoginClassifier.php [hunk 1]
```diff
@@ -105,8 +105,13 @@ public function process(string $uid, string $ip) {
...
 		$this->logger->warning("Detected a login from a suspicious login. user=$uid ip=$ip strategy=" . $strategy::getTypeName());
 
-		$login = $this->persistSuspiciousLogin($uid, $ip);
-		$this->notifyUser($uid, $ip, $login);
+		try {
+			$login = $this->persistSuspiciousLogin($uid, $ip);
+			$this->notifyUser($uid, $ip, $login);
+		} catch (\RuntimeException) {
+			// We could not persist the suspicous login, but we should still notify the user
+			$this->notifyUser($uid, $ip, null);
```
Comment (ChristophWurst) [inline lib/Service/LoginClassifier.php:113]: either log the exception or rethrow

File: lib/Service/LoginClassifier.php [hunk 2]
```diff
@@ -128,15 +133,24 @@ private function persistSuspiciousLogin(string $uid, string $ip): SuspiciousLogi
...
-	 * @param SuspiciousLogin $login
+	 * @param ?SuspiciousLogin $login
 	 */
-	private function notifyUser(string $uid, string $ip, SuspiciousLogin $login): void {
+	private function notifyUser(string $uid, string $ip, ?SuspiciousLogin $login): void {
+		if ($login === null) {
+			// There was an error persisting the login attempt, so we can not look for related events in the past.
+			// But we should still warn the user and not silently accept that attempt.
+			$event = new SuspiciousLoginEvent($uid, $ip);
+			$this->dispatcher->dispatch(SuspiciousLoginEvent::class, $event);
```
Comment (ChristophWurst) [inline lib/Service/LoginClassifier.php:150]: dispatchTyped?

File: lib/Service/ModelStore.php [hunk 3]
```diff
@@ -148,7 +148,7 @@ public function load(int $id): Estimator {
 	}
 
 	/**
-	 * @param Estimator|Persistable $estimator
+	 * @param Estimator $estimator (Must implement Persistable)
```
Comment (ChristophWurst) [inline lib/Service/ModelStore.php:151]: what about Estimator&Persistable? https://psalm.dev/docs/annotating_code/type_syntax/intersection_types/

**Review by ChristophWurst — COMMENTED** [review]
Thanks a lot

File: lib/Service/ModelStore.php (hunk 3 above)
Comment (susnux) [reply lib/Service/ModelStore.php:151]: Can be done in a follow up, but this causes many issues as all functions also need to return the correct type.

File: lib/Service/LoginClassifier.php (hunk 1 above)
Comment (susnux) [reply lib/Service/LoginClassifier.php:113]: Good point, fixed

File: lib/Service/LoginClassifier.php (hunk 2 above)
Comment (susnux) [reply lib/Service/LoginClassifier.php:150]: makes sense :)

**Review by ChristophWurst — APPROVED** [review]

**Comment by susnux** [comment]
>  php8.0 lint Expected — Waiting for status to be reported 

This needs to be removed from the expected checks - it does not exist as we can only check 8.1+

**Pull request merged**